File: LeetCode/101-200/145. 二叉树的后序遍历.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
 func postorderTraversal(root *TreeNode) (res []int) {
	node := root
var stack []*TreeNode
var prev *TreeNode
for node != nil || len(stack) > 0 {
	for node != nil {
		stack = append(stack, node)
		node = node.Left
	}
	node = stack[len(stack)-1]
	stack = stack[:len(stack)-1]
	if node.Right == nil || node.Right == prev {
		res = append(res, node.Val)
		prev = node
		node = nil
	} else {
		stack = append(stack, node)
		node = node.Right
	}
}
return
}

File: LeetCode/101-200/145. 二叉树的后序遍历.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
   public List<Integer> postorderTraversal(TreeNode root) {
            List<Integer> list= new ArrayList<>();
            dfs(root,list);
            return list;
    }
    public void dfs(TreeNode root,List<Integer> l) {
        if (root == null) return;
        dfs(root.left,l);
        dfs(root.right,l);
        l.add(root.val);
    }
}